### PR TITLE
address issue #2334

### PR DESCRIPTION
**Pull request by watashiSHUN**
address issue #2334 
"function.json"
```
{
    "scriptFile": "bin/Debug/PreCompiledFunctionSample.dll",
    "entryPoint": "PreCompiledFunctionSample.MyFunction.Run",
     ...
}
```
kudu/api/functions:
```
[{
        ...
	"script_href": "https://shunfunctionapp.scm.azurewebsites.net/api/vfs/site/wwwroot/dllTest/bin/Debug/PreCompiledFunctionSample.dll",
	"config_href": "https://shunfunctionapp.scm.azurewebsites.net/api/vfs/site/wwwroot/dllTest/function.json",
	"secrets_file_href": "https://shunfunctionapp.scm.azurewebsites.net/api/vfs/data/functions/secrets/dllTest.json",
	"href": "https://shunfunctionapp.scm.azurewebsites.net/api/functions/dllTest"
        ...
}]
```
`DeterminePrimaryScriptFile()` comes from this [commit](https://github.com/Azure/azure-webjobs-sdk-script/commit/71142175acd4a48a938c79951861492583576313).

TODO: refactor this function into a library

### CODE REVIEW

File: Kudu.Core/Functions/FunctionManager.cs [hunk 1]
```diff
@@ -266,7 +266,7 @@ private FunctionEnvelope CreateFunctionConfig(string configContent, string funct
             {
                 Name = functionName,
                 ScriptRootPathHref = FilePathToVfsUri(GetFunctionPath(functionName), isDirectory: true),
-                ScriptHref = FilePathToVfsUri(GetFunctionScriptPath(functionName, functionConfig)),
+                ScriptHref = FilePathToVfsUri(DeterminePrimaryScriptFile(functionConfig, GetFunctionPath(functionName))),
```
Comment (davidebbo) [inline Kudu.Core/Functions/FunctionManager.cs]: nit: Use `functionPath` variable since it's used in two places.

File: Kudu.Core/Functions/FunctionManager.cs [hunk 2]
```diff
@@ -275,35 +275,46 @@ private FunctionEnvelope CreateFunctionConfig(string configContent, string funct
...
             }
             else
             {
-                // First see if there is an explicit primary file indicated
-                // in config. If so use that.
-                string functionPrimary = null;
-                string scriptFileName = (string)functionConfig["scriptFile"];
-                if (!string.IsNullOrEmpty(scriptFileName))
+                string[] functionFiles = FileSystemHelpers.GetFiles(scriptDirectory,"*.*", SearchOption.TopDirectoryOnly)
+                    .Where(p => Path.GetFileName(p).ToLowerInvariant() != "function.json")
```
Comment (davidebbo) [inline Kudu.Core/Functions/FunctionManager.cs]: Use case insensitive compare instead of ToLower(). Same for couple calls below. Generally, always avoid using ToLower() for comparison purpose. Use `String.Equals(s1, s2, StringComparison.OrdinalIgnoreCase) `.

File: Kudu.Core/Functions/FunctionManager.cs [hunk 3]
```diff
@@ -313,9 +324,15 @@ internal static string DeterminePrimaryScriptFile(JObject functionConfig, string
                         Path.GetFileNameWithoutExtension(p).ToLowerInvariant() == "run" ||
                         Path.GetFileName(p).ToLowerInvariant() == "index.js");
                 }
+            }
 
-                return functionPrimary;
+            if (string.IsNullOrEmpty(functionPrimary))
+            {
+                throw new ConfigurationErrorsException("Unable to determine the primary function script. Try renaming your entry point script to 'run' (or 'index' in the case of Node), " +
+                    "or alternatively you can specify the name of the entry point script explicitly by adding a 'scriptFile' property to your function metadata.");
```
Comment (davidebbo) [inline Kudu.Core/Functions/FunctionManager.cs]: You throw various exception with descriptive messages, but don't they all get ignored in `TryGetFunctionConfigAsync`? Is so, what's the point?

Comment (davidebbo) [inline Kudu.Core/Functions/FunctionManager.cs]: I wonder if the better behavior would be to just have a null `ScriptHref` if we can't figure out the primary script. This way, they still show up in the portal. Talk to @alrod to make sure Portal can handle that. I think generally we want to be fairly lenient so that user can still open function in portal and fix issue. If you skip it, it just disappears.

File: Kudu.Core/Functions/FunctionManager.cs [hunk 4]
```diff
@@ -275,35 +275,46 @@ private FunctionEnvelope CreateFunctionConfig(string configContent, string funct
...
+            // First see if there is an explicit primary file indicated
+            // in config. If so use that.
+            string functionPrimary = null;
+            string scriptFile = (string)functionConfig["scriptFile"];
+
+            if (!string.IsNullOrEmpty(scriptFile))
             {
-                // if there is only a single file, that file is primary
-                return functionFiles[0];
+                string scriptPath = Path.Combine(scriptDirectory, scriptFile);
```
Comment (davidebbo) [inline Kudu.Core/Functions/FunctionManager.cs]: Note that in some cases, the script file will be outside the folder (e.g. `../otherfolder/foo.js`). e.g. this will happen in scenarios like @christopheranderson's funcpack.

Comment (watashiSHUN) [inline Kudu.Core/Functions/FunctionManager.cs]: var filePathRelative = @"..\emptyWeb\index.html";
            var folderPath = @"D:\test";
            var pathConcate = Path.Combine(folderPath, filePathRelative); 
            var pathFull = Path.GetFullPath(pathConcate); 
            Console.WriteLine($"{pathConcate} : {File.Exists(pathConcate)}"); // D:\test\..\emptyWeb\index.html : True
            Console.WriteLine($"{pathFull} : {File.Exists(pathFull)}"); // D:\emptyWeb\index.html : True

we are calling `Path.GetFullPath` on `functionPrimary` at the end of this function

Comment (davidebbo) [inline Kudu.Core/Functions/FunctionManager.cs]: My question is more whether the caller expects the condition. i.e. would the portal then just open this as the primary file for that function even if it's not in the folder? They may cause weird behavior with the file list. Please discuss with @alrod

File: Kudu.Core/Functions/FunctionManager.cs (hunk 3 above)
Comment (watashiSHUN) [inline Kudu.Core/Functions/FunctionManager.cs]: if we return null, exception messages will be lost (which are useful information for users, whereas currently if we give an invalid url, here's the message: `"Error:We are unable to reach your function app (Not Found). Please try again later."`), maybe we can do one of the following:
1. return the error string in `ScriptHref`
2. return null, add an error field to `functionEnvelop`
3. write the exception message to a error file in function folder, and pass its uri

I personally prefer **3**, as it requires almost no change on portal, but I don't know if it can cause function runtime/(3rd party which calls `api/functions`) to behave weirdly?

File: Kudu.Core/Functions/FunctionManager.cs (hunk 4 above)
Comment (watashiSHUN) [inline Kudu.Core/Functions/FunctionManager.cs]: portal has a file system viewer, which does not support going outside of function folder

**Comment by watashiSHUN** [comment]
@davidebbo we didn't finish our conversation before you left, should I add unit tests like these? (which will cover edge cases you mentioned in code review, they are basically testing the same thing as the [runtime tests](https://github.com/Azure/azure-webjobs-sdk-script/blob/dev/test/WebJobs.Script.Tests/ScriptHostTests.cs)), I am not sure whether I should duplicate these tests in kudu? the key test in this pull request is probably something like : [`DeterminePrimaryScriptFile_RelativeSourceFileSpecified()`](https://github.com/Azure/azure-webjobs-sdk-script/blob/dev/test/WebJobs.Script.Tests/ScriptHostTests.cs#L234-L254)

File: Kudu.Core/Functions/FunctionManager.cs [hunk 5]
```diff
@@ -275,47 +276,71 @@ private FunctionEnvelope CreateFunctionConfig(string configContent, string funct
...
+                    .ToArray();
+
+                if (functionFiles.Length == 0)
                 {
-                    functionPrimary = functionFiles.FirstOrDefault(p =>
-                        string.Compare(Path.GetFileName(p), scriptFileName, StringComparison.OrdinalIgnoreCase) == 0);
+                    TraceAndThrowError(new ConfigurationErrorsException("No function script files present."));
+                }
+
+                if (functionFiles.Length == 1)
```
Comment (ahmelsayed) [inline Kudu.Core/Functions/FunctionManager.cs]: else if?

Comment (davidebbo) [inline Kudu.Core/Functions/FunctionManager.cs]: No point, it throws
